build/builder.go

```diff
@@ -539,29 +583,100 @@ func (b *Builder) Finish() error {
 		return b.buildError
 	}
 
+	// map of temporary -> final names for all updated shards + shard metadata files
+	artifactPaths := make(map[string]string)
+	for tmp, final := range b.finishedShards {
+		artifactPaths[tmp] = final
+	}
+
+	oldShards := b.opts.FindAllShards()
+
+	if b.opts.IsDelta {
+		// Delta shard builds need to update FileTombstone and branch commit information for all
+		// existing shards
+		for _, shard := range oldShards {
+			repositories, indexMetadata, err := zoekt.ReadMetadataPathAlive(shard)
+			if err != nil {
+				b.buildError = err
+				continue
+			}
+
+			for _, r := range repositories {
+				if r.ID == b.opts.RepositoryDescription.ID {
+
+					if len(b.opts.ChangedOrRemovedFiles) > 0 && r.FileTombstones == nil {
+						r.FileTombstones = make(map[string]struct{})
+					}
+
+					for _, f := range b.opts.ChangedOrRemovedFiles {
+						r.FileTombstones[f] = struct{}{}
+					}
+
+					// TODO: Should we fail early here if we detect that the set of branch names
+					// in the builder options differs from what's inside the repository? Otherwise,
+					// the only error we'll get is a panic when we actually search the shard.
+					r.Branches = b.opts.RepositoryDescription.Branches
+
+					break
+				}
+			}
+
+			var updatedRepositories interface{}
+
+			if indexMetadata.IndexFormatVersion >= 17 {
```

Comment:
I feel like this check is a bit of a footgun (needing to remember to check the index version to know how to serialize the repository metadata file). We also need to do this in both the shard merger and now in the builder for delta shards. Perhaps we should write a shared utility for this?

Reply:
Can't we skip this entirely? The builder never builds compound shards. Merge doesn't even call it. We can savely assume that we only every write out v16 shards here, can't we?

Reply:
I added this check because I was seeing an error if one of the pre-exisitng shards was a compound shard. Perhaps it'd be better just make it fail with an error like ("can't create a delta shard for a repository inside a compound shard" for now though).

Reply:
I added the error here: https://github.com/sourcegraph/zoekt/pull/274/commits/91525b3fe8c7ba910684b2759aef4af1ce91a74e . I was able to get rid of the for loop logic as a side effect.

Reply:
Maybe we could we check in IndexState() and only trigger a delta build if the repo was not inside a compound shard?